src/mon/command_multiplexer.h

```diff
@@ -0,0 +1,312 @@
+    * register the type of your command with the registry helper
+
+In practice, a minimal function would look something like this:
+
+    struct my_command : public my_command_t
+    {
+        vector<string> tags() { return { "my_command" }; }
+    };
+
+...next, you need to register the type of your command function. The framework deployment will probably
+provide a place for you to do this, most likely a free function called register_commands(). There
+will be a place for you to add your command type. Below, we've added "my_command":
+
+    void register_commands(my_command_registry& commands)
+    {
+     ceph::register_commands<my_command_registry, 
+        my_command
+     >(commands);
+    }
+
+That will automatically register your function, set up the command bindings, set up multiplexing for
+whatever phase (preprocess, prepare) you've defined, and handle interpreting the results of your function,
+leaving you to concentrate on what you want it to do.
+
+Next, you'll probably want to define member functions for only the phases (ie. prepare, preprocess) that you
+wish to respond to. For example, we can extend my_command with a prepare command like so:
+
+    struct my_command : public my_command_t
+    {
+        vector<string> tags() { return { "my_command" }; }
+
+        my_command_result at_prepare(const my_command_ctx& state, string_view cmd);
+    };
+
+...that's it. 
+
+Your function's exact call is in the "cmd" parameter, and any other state is communicated via the
+"state" parameter (which you may call something shorter if you like); the exact content depends on
+what is available in your particular deployment.
+
+Next, let's consider what to return from our functions. This is defined by your local deployment, there 
+is no fixed type. Most of the time, you'll probably see something like this:
+    using my_cmd_result = std::tuple<my_cmd_status, int>;
+
+Where my_cmd_status is something like:
+
+    enum class my_cmd_status : int { failure = 0, success };
+
+...but, once again, nothing in the mini-framework enforces this. Above, the enum is allowed to degrade
+to int because the functions it wraps expects an int <= 0 on failure, > 0 on success back.
+
+Now, our final example:
+
+    struct my_command : public my_command_t
+    {
+        vector<string> tags() { return { "my_command" }; }
+
+        my_command_result at_prepare(const my_command_ctx& state, string_view cmd)
+        {
+            return { my_cmd_status::success, 1 };
+        }
+    };
+
+ Design note on context structure:
+	Any similar design has to have some way of passing parameters into individual commands. In
+ the original version, commands were implemented as member functions within the base class, and
+ has these implicitly available. By separating them, we're left with several choices: on one hand,
+ we could pass them in the signature for each command. This has the advantage of not requiring a
```

Comment:
This is an important design point worth discussing. 

First, it's /definitely/ worthwhile (to me anyway!) to have a goal of getting command definitions out of these command function objects themselves-- it may be easier with some of the C++17 introspection tools, however it won't be trivial and above all it would create more load on the /user/ of this library: they'd wind up having to add things like the output ostream& all the time, etc. and deal once again with having to have very specific knowledge of each parameter (more below). BUT I'll maybe tinker with at least getting an understanding of what it would take to get some kind of way to specify what getting the signatures into the classes. One start might very "simply" be having a const char[] with the text inside the class (specific field name) that can be bound by the registration function. OTOH, I'm worried this might cause extra implementation churn and/or run into trouble with the COMMAND and COMMAND_WITH_FLAG macros, and things like FLAG(NOFORWARD). The complexity unfortunately probably warrants another separate pass, with its own considerations IMO. Fortunately, it's something that can be ADDED, I don't think there's anything stopping us from doing it in the (hopefully near) future.

With respect to using a context structure:
Three of the main goals for this framework are to make adding new commands "easy", to see which values are constant or modified by side-effect, and to provide consistent names. The approach of using a context object tries to resolve this in a few ways:
1a) signatures are concise to write: user of framework has little to remember
1b) important corollary: changing the content of the context (i.e. adding or removing a member) does not effect any function signatures, no need to update every single function
2) a single place exists to describe const-ness of members;
3a) a single place exists to name members;
3b) a single place to /look up/ what's available
3c) context structure's ctor has already taken care of parameter-fetching boilerplate 

Right now, commands have full access to private members of Monitor, which means they can have hidden "parameters" and modify state at a distance. They also do so, and it requires careful inspection of the code (in an 8k line function) to really see what does what and where.

Though not perfect, the context approach avoids these situations (but has its own exchanges, to be sure). I somewhat /like/ the idea that there's a slight amount of extra typing "with the fingers" to interact with the outside-world data in the structure ("ctx.ss" vs. "ss") and especially given that references are free I haven't noticed in my porting experiments that it's a significant barrier.

Another thought: there's lots of committee work on structure introspection, and we could /also/ choose to use a tuple and reference members to do more compile-time magic on the context if we want to. But I'm feeling enough complexity in this initial version to not make me want to do that if there's no reason to.